src/util/result.h

```diff
@@ -57,15 +63,77 @@ namespace util {
 //! return function results.
 //!
 //! Usage examples can be found in \example ../test/result_tests.cpp.
-template <typename SuccessType = void, typename FailureType = void, typename MessagesType = bilingual_str>
+template <typename SuccessType = void, typename FailureType = void, typename MessagesType = Messages>
 class Result;
 
 //! Wrapper object to pass an error string to the Result constructor.
 struct Error {
     bilingual_str message;
 };
+//! Wrapper object to pass a warning string to the Result constructor.
+struct Warning {
+    bilingual_str message;
+};
+
+//! Wrapper object to pass an existing Result object to the Result constructor,
+//! moving messages into the new object, like operator>> below. Passing
+//! MoveFrom(result) to the Result constructor is similar to passing
+//! std::move(result), except it doesn't require the two result objects to have
+//! compatible Success and Failure types, and it only moves message and info
+//! fields, leaving success and value fields alone. Like operator>>, it is
+//! useful for combining data from many result objects into a single result.
+template <typename Result>
+struct MoveFrom {
+    MoveFrom(Result&& result) : m_result(result) {}
+    Result& m_result;
+};
+
+//! Template deduction guide for MoveFrom class.
+template <class Result>
+MoveFrom(Result&& result) -> MoveFrom<Result>;
+
+//! Extension point for overriding the way SuccessType / FailureType /
+//! MergesType values are combined.
+template<typename T>
+struct ResultTraits {
+    template<typename O>
+    static void MergeInto(O& dst, T& src)
+    {
+        dst = std::move(src);
+    }
+};
+
+//! Extension point for specializing behavior of MessagesType
+template<typename MessagesType>
+struct MessagesTraits;
+
+//! ResultTraits specialization for Messages struct.
+template<>
+struct ResultTraits<Messages> {
+    static void MergeInto(Messages& dst, Messages& src);
+};
+
+//! MessagesTraits specialization for Messages struct.
+template<>
+struct MessagesTraits<Messages> {
```

Comment:
I have not quite grasped yet what the immediate benefit of these more generic decorators on the Messages are. Am I missing something from your draft PRs? I like shiny generics though, so to me a clear benefit of this approach would be to future-proof the result type. I was asking myself if this could be a bit more readable if instead of a forward-declared struct, the `MessagesTraits` were made a concept. After implementing it, I think it is indeed a bit easier to parse at the call site. It also potentially allows the user to define their own types for the warnings and errors fields like this: https://github.com/TheCharlatan/bitcoin/commit/f5e5442cf80c58c0509562d8c689243c8d9becf2 . Could this also be leveraged in place of your future addition of an info type?